File: Dataset/Leetcode/train/22/258.py
 class Solution:
    def XXX(self, n: int) -> List[str]:
        res = []
        temp = []
        def backtrack(ln,rn):
            if not ln and not rn:
                res.append(''.join(temp))
            if ln:
                temp.append('(')
                backtrack(ln -1,rn) # 递归 左括号
                temp.pop() # 回溯
            if rn > ln: # ln小于rn时，添加")"可以构成完整括号
                temp.append(')')
                backtrack(ln,rn-1) # 递归 右括号
                temp.pop() # 回溯
        if not n:
            return res
        backtrack(n,n)
        return res

File: Dataset/Leetcode/train/20/258.cpp
 class Solution {
public:
    bool XXX(string s) {
	map<char, int>mymap;
	mymap['('] = 0;
	mymap['{'] = 1;
	mymap['['] = 2;
	mymap[')'] = 3;
	mymap['}'] = 4;
	mymap[']'] = 5;
	stack<char>mys;
	for (int i = 0; i < s.size(); i++)
	{
		if (isspace(s[i]))
			continue;
		if (mymap[s[i]] <= 2)
			mys.push(s[i]);
        else if(mys.empty()&&mymap[s[i]] >2)
            return false;
		else if (!mys.empty() && (mymap[s[i]] - mymap[mys.top()] != 3))
			return false;
		else
			mys.pop();
	}
	return mys.empty();
}
};

